Firestore/core/src/firebase/firestore/model/field_transform.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_TRANSFORM_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_TRANSFORM_H_
+
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/model/field_path.h"
+#include "Firestore/core/src/firebase/firestore/model/transform_operations.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/** A field path and the TransformOperation to perform upon it. */
+class FieldTransform {
+ public:
+  FieldTransform(FieldPath path, std::unique_ptr<TransformOperation> transform)
+      : path_(std::move(path)), transform_(std::move(transform)) {
+  }
+
+  const FieldPath& path() const {
+    return path_;
+  }
+
+  const TransformOperation* transform() const {
+    return transform_.get();
+  }
+
+#if defined(__OBJC__)
+  bool operator==(const FieldTransform& other) const {
+    return path_ == other.path_ && *transform_ == *other.transform_;
+  }
+
+  // For Objective-C++ hash; to be removed after migration.
+  // Do NOT use in C++ code.
+  NSUInteger Hash() const {
+    NSUInteger hash = path_.Hash();
+    hash = hash * 31 + transform_->Hash();
+    return hash;
+  }
+#endif  // defined(__OBJC__)
+
+ private:
+  FieldPath path_;
+  // Shared by copies of the same FieldTransform.
+  std::shared_ptr<TransformOperation> transform_;
```

Comment:
Can it be `<const TransformOperation>`? Also, can you please provide more context for this optimization?

Reply:
Change to `const`. Really does not matter for the only `ServerTimestampTransform`. I think we can safely assume a `TransformOperation` will not mutate itself when proceeding fields.

We need to access a `TransformOperation` here. Since it is poly, we need either a pointer or a reference. Conceptually, `FieldTransform` should own it instead of each `TransformOperation` type owns it (for ServingTimestampTransform might be OK since it builds a singleton). Now if you wonder why `unique_ptr` does not work here, that's because a `FieldTransform` object can be copied.